rfcs/0004-policy-sync-v1.md

```diff
@@ -0,0 +1,182 @@
+[guide-level-explanation]: #guide-level-explanation
+
+The policy sync feature will allow a gateway-admin to configure, via GatewayClass
+parameters, a set of Policy GVRs to be synced by the Multicluster Gateway Controller.
+
+The `policiesToSync` field in the parameters defines those GVRs. For example, in
+order to configure the controller to sync AuthPolicies:
+
+```json
+"policiesToSync": [
+  {
+    "group": "kuadrant.io",
+    "version": "v1beta1",
+    "resource": "authpolicies" 
+  }
+]
+```
+
+The support for resources that the controller can sync is limited by the following:
+* The controller ServiceAccount must have permission to watch, list, and get the
+  resource to be synced
+* The resource must implement the Policy schema:
+    * Have a `.spec.targetRef` field
+
+When a Policy is configured to be synced in a GatewayClass, the Multicluster
+Gateway Controller starts watching events on the resources, and propagates changes
+by placing the policy in the spoke clusters, with the following mutations:
+* The `TargetRef` of the policy is changed to reference the downstream Gateway
+* The `kuadrant.io/policy-synced` annotation is set
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+### Process overview
+
+#### Dynamic Policy watches
+
+The Multicluster Gateway Controller reconciles parameters referenced by the
+GatewayClass of a Gateway. A new field is added to the parameters that allows
+the configuration of a set of GVRs of Policies to be synced.
+
+The GatewayClass reconciler validates that:
+* The GVRs reference existing resource definitions
+* The GVRs reference resources that implement the Policy schema.
+
+Validation failures are reported as part of the status of the GatewayClass
+
+The Gateway reconciler sets up dynamic watches to react to events on the configured
+Policies, calling the PolicySyncer component with the updated Policy as well
+as the associated Gateway.
+
+#### PolicySyncer component
+
+The PolicySyncer component is in charge of reconciling Policy watch events to
+apply the necessary changes and place the Policies in the spoke clusters.
+
+This component is injected in the event source and called when a change is made
+to a hub Policy that has been configured to be synced.
+
+The PolicySyncer implementation uses OCM ManifestWorks to place the policies in
+the spoke clusters. Through the ManifestWorks, OCM allows to:
+* Place the Policy in each spoke cluster
+* Report the desired status back to the hub using JSON feedback rules
+
+### Policy Hierarchy 
+
+In order to avoid conflict with Policies created directly in the spoke clusters,
+a hierarchy must be defined to prioritise those Policies.
+
+The controller will set the `kuadrant.io/policy-synced` annotation on the policy
+when placing it in the spoke cluster.
+
+The Kuadrant operator will be aware of the presence of this annotation, and, in case
+of conflicts, override Policies that contain this annotation.
```

Comment:
Will this be reported back to the original policy? I worry about traceability of the override

Reply:
That's a good idea, it could be reported with the `Overriden` status condition proposed in the RFC (https://github.com/Kuadrant/architecture/blob/main/rfcs/0004-policy-status.md#conditions) I'll add it to the document